cip-ledger-state.md

```diff
@@ -0,0 +1,111 @@
+---
+CIP: ?
+Title: Extended Local Chain Sync Protocol
+Authors: Erik de Castro Lopo <erikd@mega-nerd.com>
+Discussions-To: Erik de Castro Lopo <erikd@mega-nerd.com>
+Comments-Summary: Extend Local Chain Sync Protocol
+Comments-URI:
+Status: Draft
+Type: ?
+Created: 2022-11-15
+License: CC-BY-4.0
+---
+## Abstract
+
+Modify the `cardano-node` (and underlying code) to provide an extended version of the existing
+local chain sync protocol.
+
+## Motivation
+
+Applications that provide insight into the Cardano block chain (like db-sync, exporters, Kupo, and
+smart contracts reacting to events) often need access to the current state of the Cardano ledger
+(stake distribution, reward and wallet balances etc). This information is known to the node, partly
+on the block chains and partly in what is referred to as ledger state (described more fully below).
+Extracting block chain data is relatively easy, but ledger state data is not. Currently these
+applications have to recreate and maintain ledger state themselves based on the block information
+they stream from the node over the local chain sync protocol. Recreation and maintenance of the
+ledger state is not only complex and a source of bugs but more importantly requires significant
+resources, specifically RAM. Ledger state in memory currently consumes 10 Gig of RAM and that is
+growing. In situations where the node and any such application run on the same machine, the machine
+ends up with twice the resource usage. The following proposal hopes to reduce resource usage and
+complexity for chain following applications like db-sync.
+
+# Current Situation
+
+Currently there is a local chain sync protocol which is really just the peer-to-peer protocol
+using a local domain socket rather than the TCP/IP socket normally used for P2P transport.
+
+The data transported over this local chain sync protocol is limited to block chain data. However, a
+Cardano node also maintains ledger state which includes:
+
+* The current UTxO state.
+* Current amount of ADA delegated to each stake pool.
+* Which stake address is currently delegated to each pool.
+* Rewards account balances for each stake address.
+* Current protocol parameters.
+
+The first of these ledger state components is by far the largest component and is probably not
+needed outside the node (and definitely not needed by db-sync). However the others are needed and
+stored by `cardano-db-sync` which gets these data sets by maintaining its own copy of ledger state
+and periodically extracting the parts required.
+
+This means that when `node` and `db-sync` are run on the same machine (which is the recommended
+configuration) that machine has two basically identical copies of ledger state. Ledger state is a
```

Comment:
I think there's little use-case for the UTxO state or the protocol parameters because these are straightforward to obtain from the chain-sync protocol itself. 

However, anything that regards _rewards_ is indeed a pain in the *** to work with; because rewards are _implicit_ in the protocol. I'd love to see a quick justification here (instead of "basically" 😅 ...) stating just that. There's no way for a client application to provide reliable information on rewards without maintaining *the entire ledger state*; for one needs to know what everyone owns in order to calculate rewards snapshots on each era.

Reply:
I am pretty sure that the protocol parameters are *not* available on chain, only the voting for the changes. Yes, the ledger rules can be applied to whatever voting is seen on chain but that is way more fragile than getting the parameters from the source.

Agree, that nobody in their right mind would want the UTxO state.

Reply:
the protocol parameters are available on chain, `NewEpochState` -> `EpochState` -> `PParams`.

Reply:
It would be really nice to have a function to pull the protocol parameters out of ledger state so that I did not have to go digging around inside it.

Reply:
you're in luck, I have one right here: `nesEs . esPp`

Reply:
I would call that "digging into the internals".

Reply:
The lack of a quality api is indeed an issue, but a much bigger issues imo, is the need to replicate resources outside the node, especially for ledger state data. An api cannot help there, this requires improvements on the protocol level. Api and protocol are two different topics.